### PR TITLE
limine: init at 7.3.0

### PR DESCRIPTION
**Pull request by phip1611**
## Description of changes

This inits the Limine bootloader. Unlike https://github.com/NixOS/nixpkgs/pull/194841, this only inits the package for now. The package itself is very useful already, for example for OS developers that want to boot their kernel using Limine. (Similar to using `grub-mkstandalone` or `grub-mkrescue`)

Adding `boot.loader.limine.enable` as NixOS option will be a follow-up.

## Things done

- Built on platform(s)
  - [ ] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- For non-Linux: Is sandboxing enabled in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
  - [ ] `sandbox = relaxed`
  - [ ] `sandbox = true`
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [24.05 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2405.section.md) (or backporting [23.05](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2305.section.md) and [23.11](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2311.section.md) Release notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [ ] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

---

Add a :+1: [reaction] to [pull requests you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[pull requests you find important]: https://github.com/NixOS/nixpkgs/pulls?q=is%3Aopen+sort%3Areactions-%2B1-desc

### CODE REVIEW

**Comment by phip1611** [comment]
Ping @blitz, @astro, and @48cf for review

File: pkgs/by-name/li/limine/default.nix [hunk 1]
```diff
@@ -0,0 +1,61 @@
...
+  ];
+  doCheck = false;
+  configurePhase = ''
+    ./configure --enable-all
+  '';
+  buildPhase = ''
+    make -j $(nproc)
+  '';
+  installPhase = ''
+    cp -R bin/. $out
```
Comment (phip1611) [inline pkgs/by-name/li/limine/default.nix:43]: We should use `destdir=xyz make install`

see https://github.com/limine-bootloader/limine/pull/340#discussion_r1535607767

File: pkgs/by-name/li/limine/package.nix [hunk 2]
```diff
@@ -0,0 +1,61 @@
...
+, nasm
+}:
+
+let
+  version = "7.1.0";
+  # It is recommended by limine to use the release tarballs instead the git repo.
+  limineSrc = fetchurl {
+    url = "https://github.com/limine-bootloader/limine/releases/download/v${version}/limine-${version}.tar.gz";
+    sha256 = "sha256-oaQA461MGVuG7aJri2HyYb6Pu8BXOWTF3Wq/SputQSw=";
+  };
```
Comment (astro) [inline pkgs/by-name/li/limine/package.nix:21]: `limineSrc` is used only once. Inline it!

File: pkgs/by-name/li/limine/package.nix [hunk 3]
```diff
@@ -0,0 +1,61 @@
...
+in
+stdenvNoCC.mkDerivation {
```
Comment (astro) [inline pkgs/by-name/li/limine/package.nix:23]: Make this attrset `rec` so that `version` becomes accessible to `src`.

File: pkgs/by-name/li/limine/package.nix [hunk 4]
```diff
@@ -0,0 +1,61 @@
...
+    llvmPackages.bintools
+    llvmPackages.clang
+    llvmPackages.lld
+    mtools
+    nasm
+  ];
+  doCheck = false;
+  configurePhase = ''
+    ./configure --enable-all
+  '';
```
Comment (astro) [inline pkgs/by-name/li/limine/package.nix:38]:
```suggestion
  configureFlags = [ "--enable-all" ];
```

File: pkgs/by-name/li/limine/package.nix [hunk 5]
```diff
@@ -0,0 +1,61 @@
...
+  buildPhase = ''
+    make -j $(nproc)
+  '';
```
Comment (astro) [inline pkgs/by-name/li/limine/package.nix:41]: No explicit `buildPhase` needed

File: pkgs/by-name/li/limine/package.nix [hunk 6]
```diff
@@ -0,0 +1,61 @@
...
+  installPhase = ''
+    cp -R bin/. $out
+  '';
```
Comment (astro) [inline pkgs/by-name/li/limine/package.nix:44]: The `/.` can be left out.

If the `bin` subdir shall be kept:
```suggestion
  installPhase = ''
    mkdir $out
    cp -R bin $out/
  '';
```

**Review by astro — CHANGES_REQUESTED** [review]

File: pkgs/by-name/li/limine/package.nix (hunk 4 above)
Comment (phip1611) [reply pkgs/by-name/li/limine/package.nix:38]: Neat, thanks!

File: pkgs/by-name/li/limine/package.nix (hunk 3 above)
Comment (blitz) [reply pkgs/by-name/li/limine/package.nix:23]: very-nit: At least nix.dev [recommends against rec](https://nix.dev/guides/best-practices#recursive-attribute-set-rec). I've seen this pattern a lot recently:

```nix
stdenv.mkDerivation (finalAttrs: {
  src = fetchFromGitHub {
    owner = "falkTX";
    repo = finalAttrs.pname;
    rev = "v${finalAttrs.version}";
    hash = "sha256-H15T/z/IRfgWdqToTzq2eJ7q3n9Kj44IZXsd4uaipuU=";
  };
  # ...
```

Not sure this is much better. The simple `let` is pretty ok in my view, because it's easy to see what happens (also when overriding it).

File: pkgs/by-name/li/limine/package.nix [hunk 7]
```diff
@@ -0,0 +1,61 @@
...
+  src = limineSrc;
+  nativeBuildInputs = [
+    gnumake
+    llvmPackages.bintools
+    llvmPackages.clang
+    llvmPackages.lld
+    mtools
+    nasm
+  ];
+  doCheck = false;
```
Comment (blitz) [inline pkgs/by-name/li/limine/package.nix:35]: Can you add a comment that says why you disable it?

very nit: Grouping the attributes with some empty lines would really help with readability. Check other deriviation files for inspiration.

File: pkgs/by-name/li/limine/package.nix [hunk 8]
```diff
@@ -0,0 +1,61 @@
+# Builds limine with all available features.
+
+{
+  # Helpers
+  fetchurl
+, lib
+, stdenvNoCC
+, # Dependencies
+  llvmPackages
```
Comment (blitz) [inline pkgs/by-name/li/limine/package.nix:9]: Does `clangStdenv` work for building limine instead using `stdenvNoCC` and adding clang/LLVM?

File: pkgs/by-name/li/limine/package.nix (hunk 6 above)
Comment (blitz) [inline pkgs/by-name/li/limine/package.nix:44]: Have you tried the default `install` target? There Makefile seems to do the right thing. This would also install man pages and development headers.

The `bin/limine` tool seems to refer to the EFI files anyhow. That means they also need to be installed (which is what the normal install target does).

tl;dr

Remove the manual `installPhase`. 

Add `outputs = [ "out" "doc" "dev" "man" ];` to only install what people need.

File: pkgs/by-name/li/limine/package.nix (hunk 8 above)
Comment (blitz) [reply pkgs/by-name/li/limine/package.nix:9]: For bootloaders you may need to play with `hardeningDisable`, if compiler-hardening related issues pop up.

**Review by blitz — CHANGES_REQUESTED** [review]
So with this packaged in nixpkgs, how would anyone use it? Is there still the plan of having this as a bootloader for NixOS or can this be used without the module?

**Comment by phip1611** [comment]
> So with this packaged in nixpkgs, how would anyone use it? Is there still the plan of having this as a bootloader for NixOS or can this be used without the module?

Yes, it is useful. One can create (hybrid)bootable images using Limine as bootloader with the tools build in the derivation. Similar to what `grub-mkrescue` or `grub-mkstandalone` can do

File: pkgs/by-name/li/limine/package.nix (hunk 6 above)
Comment (phip1611) [reply pkgs/by-name/li/limine/package.nix]: > outputs = [ "out" "doc" "dev" "man" ];

love it, thanks!

File: pkgs/by-name/li/limine/package.nix (hunk 8 above)
Comment (phip1611) [reply pkgs/by-name/li/limine/package.nix:9]: It builds without errors, so when should this be a problem?

**Comment by phip1611** [comment]
Ping @SuperSandro2000

File: pkgs/by-name/li/limine/package.nix (hunk 8 above)
Comment (blitz) [reply pkgs/by-name/li/limine/package.nix:9]: Some low-level software is picky when certain features are enabled, because it violates their assumptions. Baremetal code is only compatible with `-fstack-protector` when it provides the right symbols, for example.

**Review by blitz — APPROVED** [review]

File: pkgs/by-name/li/limine/package.nix [hunk 9]
```diff
@@ -0,0 +1,59 @@
...
+    # Caution. Some submodules have different licenses.
+    license = licenses.bsd2;
+    platforms = [
+      "i686-linux"
+      "x86_64-linux"
+      "x86_64-darwin"
+      "aarch64-linux"
+      "aarch64-darwin"
+      "riscv-linux"
+    ];
```
Comment (SuperSandro2000) [inline pkgs/by-name/li/limine/package.nix:53]: Isn't this just platforms.linux?

File: pkgs/by-name/li/limine/package.nix [hunk 10]
```diff
@@ -0,0 +1,59 @@
...
+}:
+
+let
+  version = "7.1.0";
+in
+clangStdenv.mkDerivation {
+  inherit version;
+  pname = "limine";
+  src = fetchurl {
+    url = "https://github.com/limine-bootloader/limine/releases/download/v${version}/limine-${version}.tar.gz";
```
Comment (SuperSandro2000) [inline pkgs/by-name/li/limine/package.nix:21]: Why do we want to use the release asset instead of the downloading the source code?

Comment (phip1611) [reply pkgs/by-name/li/limine/package.nix:26]: We don't use the Git source but the release tarball, as the source has a
`./bootstrap` script performing non-standard network access to download
resources (a custom Git submodule-like replacement). Packaging this in Nix
is very cumbersome compared to using the release tarball, where the
`./bootstrap` step was already executed, and we can `./configure` right away.
  

I added a comment to reflect that

File: pkgs/by-name/li/limine/package.nix (hunk 9 above)
Comment (phip1611) [reply pkgs/by-name/li/limine/package.nix]: I'm not exactly sure what to put here. I think, by using clang, this should (cross)compile on every architecture properly. However, the resulting files may not run on all platforms.

@48cf @mintsuki what can you say about the platform support of Limine?

Comment (phip1611) [reply pkgs/by-name/li/limine/package.nix]: After an offline discussion with @blitz: We should use the intersection between platforms where Limine builds and where the userspace tools are useful. So I think

`platforms.linux` is correct. Possibly also `platforms.darwin`. Aye?

File: pkgs/by-name/li/limine/package.nix (hunk 8 above)
Comment (mintsuki) [reply pkgs/by-name/li/limine/package.nix:9]: I would suggest using `stdenv` as normal to build the `limine` executable with the default `cc`. This does not affect the compiler used to compile the actual bootloader, which is controlled by `CC_FOR_TARGET` and will default on `clang` if present.

File: pkgs/by-name/li/limine/package.nix [hunk 11]
```diff
@@ -0,0 +1,60 @@
...
+in
+clangStdenv.mkDerivation {
+  inherit version;
+  pname = "limine";
+  # We don't use the Git source but the release tarball, as the source has a
+  # `./bootstrap` script performing non-standard network access to download
+  # resources (a custom Git submodule-like replacement). Packaging this in Nix
+  # is very cumbersome compared to using the release tarball, where the
+  # `./bootstrap` step was already executed and we can `./configure` right
+  # away.
```
Comment (mintsuki) [inline pkgs/by-name/li/limine/package.nix:25]: My personal opinion is that this comment is overly noisy... The standard way to build autotools software as an end user is to fetch the release tarball... this is not unusual. Just like it is not unusual for `./bootstrap` or `./autogen.sh` to perform extra duties that may involve network access.

File: pkgs/by-name/li/limine/package.nix (hunk 10 above)
Comment (mintsuki) [reply pkgs/by-name/li/limine/package.nix:27]: See https://github.com/NixOS/nixpkgs/pull/298048/files#r1539841844

File: pkgs/by-name/li/limine/package.nix [hunk 12]
```diff
@@ -0,0 +1,60 @@
...
+  ];
+
+  configureFlags = [
+    "TOOLCHAIN_FOR_TARGET=llvm"
+    "--enable-all"
+  ];
+
+  installFlags = [ "destdir=$out" "manprefix=/share" ];
+
+  outputs = [ "out" "doc" "dev" "man" ];
```
Comment (mintsuki) [inline pkgs/by-name/li/limine/package.nix:45]: Are these `installFlags` and `outputs` options really necessary? Judging from looking at other autotools packages, they do not seem to be, and they seem to potentially be in error.

File: pkgs/by-name/li/limine/package.nix [hunk 13]
```diff
@@ -0,0 +1,60 @@
...
+  llvmPackages
+, mtools
+, nasm
+}:
+
+let
+  version = "7.1.0";
+in
+clangStdenv.mkDerivation {
+  inherit version;
```
Comment (mintsuki) [inline pkgs/by-name/li/limine/package.nix:18]: Couldn't this simply be moved inside the `mkDerivation` block as `version = "7.1.0";`?

File: pkgs/by-name/li/limine/package.nix [hunk 14]
```diff
@@ -0,0 +1,60 @@
...
+  outputs = [ "out" "doc" "dev" "man" ];
+
+  meta = with lib; {
+    homepage = "https://limine-bootloader.org/";
+    description = "Limine Bootloader";
+    # Caution. Some submodules have different licenses.
+    license = licenses.bsd2;
+    # While this may (cross)build on all platforms, Limine can't necessarily
+    # create bootable images for all of those platforms.
+    platforms = platforms.all;
```
Comment (mintsuki) [inline pkgs/by-name/li/limine/package.nix:54]: Maybe `platforms.unix` instead?

**Comment by mintsuki** [comment]
@phip1611 I left a review of most of the (perceived) issues I see with this PR.

File: pkgs/by-name/li/limine/package.nix (hunk 11 above)
Comment (phip1611) [reply pkgs/by-name/li/limine/package.nix]: Thanks! I changed the doc. Hope the new version is better

File: pkgs/by-name/li/limine/package.nix (hunk 13 above)
Comment (phip1611) [reply pkgs/by-name/li/limine/package.nix:18]: then we would need `rec` which should not be used:

https://github.com/NixOS/nixpkgs/pull/298048#discussion_r1535636936

**Comment by phip1611** [comment]
> @phip1611 I left a review of most of the (perceived) issues I see with this PR.

thanks. I think this is ready for nixpkgs. @astro can you help us to get this merged, please?

File: pkgs/by-name/li/limine/package.nix (hunk 8 above)
Comment (phip1611) [reply pkgs/by-name/li/limine/package.nix:9]: done, works

File: pkgs/by-name/li/limine/package.nix (hunk 12 above)
Comment (mintsuki) [reply pkgs/by-name/li/limine/package.nix:45]: Hmm okay. But *why* are they necessary over letting Nix do its thing? Is this something that changed compared to some older Nix version? What happens differently if these 2 assignments are omitted?

Pardon my insistence, I am asking because I have checked against several autotools packages already packaged in nixpkgs and they don't have this.

File: pkgs/by-name/li/limine/package.nix (hunk 13 above)
Comment (mintsuki) [reply pkgs/by-name/li/limine/package.nix:20]: Alright :+1:

File: pkgs/by-name/li/limine/package.nix [hunk 15]
```diff
@@ -0,0 +1,60 @@
...
+  installFlags = [ "destdir=$out" "manprefix=/share" ];
+
+  outputs = [ "out" "doc" "dev" "man" ];
+
+  meta = with lib; {
+    homepage = "https://limine-bootloader.org/";
+    description = "Limine Bootloader";
+    # Caution. Some submodules have different licenses.
+    license = licenses.bsd2;
+    # The platforms on that the Liminine binary and helper tools can run, not
```
Comment (mintsuki) [inline pkgs/by-name/li/limine/package.nix:52]: Spelling :^)

**Comment by mintsuki** [comment]
I saw that other packages have an `enableParallelBuilding = true;` assignment; perhaps it makes sense to add that here as well?

File: pkgs/by-name/li/limine/package.nix (hunk 12 above)
Comment (phip1611) [reply pkgs/by-name/li/limine/package.nix:47]: I hope this helps @mintsuki 

image

**Comment by phip1611** [comment]
> I saw that other packages have an `enableParallelBuilding = true;` assignment; perhaps it makes sense to add that here as well?

Good catch! I thought it is the default but apparently, it is not.

File: pkgs/by-name/li/limine/package.nix (hunk 12 above)
Comment (mintsuki) [reply pkgs/by-name/li/limine/package.nix:47]: Alright, okay about the `outputs` line, but what about `installFlags`? Limine's build system never uses `destdir` nor `manprefix` variables... are you sure this is doing anything? The only precedent I can find for this is in the `pkgs/os-specific/linux/mceinject/default.nix` package, but in that case the plain makefile mce-inject uses does use `destdir` and `manprefix`.

File: pkgs/by-name/li/limine/package.nix [hunk 16]
```diff
@@ -0,0 +1,62 @@
...
+    llvmPackages.lld
+    mtools
+    nasm
+  ];
+
+  configureFlags = [ "--enable-all" ];
+
+  enableParallelBuilding = true;
+  # the bootloader build takes care of the relevant flags by itself
+  NIX_HARDENING_ENABLE = "";
```
Comment (mintsuki) [inline pkgs/by-name/li/limine/package.nix:43]: Is this necessary? You know, you probably don't want to touch hardening flags that affect the normal `limine` executable that runs on the host... The bootloader itself uses `CFLAGS_FOR_TARGET` and `LDFLAGS_FOR_TARGET` and, as you said, it takes care of most flags necessities by itself.

File: pkgs/by-name/li/limine/package.nix (hunk 12 above)
Comment (phip1611) [reply pkgs/by-name/li/limine/package.nix:47]: I have that from @blitz - Can you help out?

Comment (phip1611) [reply pkgs/by-name/li/limine/package.nix:43]: Okay, `installFlags` can be dropped. The Nix magic still work. But we need `outputs` tho to get everything right.

File: pkgs/by-name/li/limine/package.nix (hunk 16 above)
Comment (phip1611) [reply pkgs/by-name/li/limine/package.nix]: This was more a preliminary measurement. Probably unneeded. Removed it.

**Comment by mintsuki** [comment]
LGTM now, thanks!

**Review by Mintsuki — APPROVED** [review]

**Comment by phip1611** [comment]
Maybe @Aleksanaa @fabaff have some resources to approve and merge?

**Review by astro — CHANGES_REQUESTED** [review]

File: pkgs/by-name/li/limine/package.nix [hunk 17]
```diff
@@ -0,0 +1,58 @@
...
+    description = "Limine Bootloader and Tools";
+    # Caution. Some submodules have different licenses.
+    license = licenses.bsd2;
+    # The platforms on that the Limine binary (helper tools) can run, not
+    # necessarily the platforms for that bootable images can be created.
+    platforms = platforms.unix;
+    maintainers = [
+      maintainers._48cf
+      maintainers.phip1611
+    ];
```
[the remark this answers is not in the document]
Comment (phip1611) [reply pkgs/by-name/li/limine/package.nix:56]: Can you please explain why you prefer this style?

Personally, and many other (non Nix) projects prefer multiline because this
way you don't pollute `git blame` when adding or removing something

File: pkgs/by-name/li/limine/package.nix [hunk 18]
```diff
@@ -0,0 +1,58 @@
...
+
+  enableParallelBuilding = true;
+
+  outputs = [ "out" "doc" "dev" "man" ];
+
+  meta = with lib; {
+    homepage = "https://limine-bootloader.org/";
+    description = "Limine Bootloader and Tools";
+    # Caution. Some submodules have different licenses.
+    license = licenses.bsd2;
```
[the remark this answers is not in the document]
Comment (phip1611) [reply pkgs/by-name/li/limine/package.nix:49]: The maintainer said it is okay to only expose this single license.

[the remark this answers is not in the document]
Comment (mintsuki) [reply pkgs/by-name/li/limine/package.nix:49]: Well, we didn't really say that, we are not lawyers... If it's better to be explicit and list the licenses of all submodules so be it. I guess it would be worth looking for similar precedents in nixpkgs to see if all submodule licenses get exposed or not.

File: pkgs/by-name/li/limine/package.nix (hunk 17 above)
[the remark this answers is not in the document]
Comment (astro) [reply pkgs/by-name/li/limine/package.nix:64]: Multiline is ok. I just suggested to reduce the `maintainers` repetition which isn't too fatal at two occurrences.
